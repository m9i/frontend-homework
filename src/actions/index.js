export const SHOW_DROPDOWN = "SHOW_DROPDOWN";
export const SET_COMPANY = "SET_COMPANY";
export const SELECT_COMPANY = "SELECT_COMPANY";
