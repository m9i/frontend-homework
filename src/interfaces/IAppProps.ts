export interface IAppProps {
    showDropdown: any;
    hideDropdown: any;
}
