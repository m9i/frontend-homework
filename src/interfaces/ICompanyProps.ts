import {ICompany} from './ICompany';

export interface ICompanyProps {
    onCompanyChange: any;
    selectedCompany: ICompany;
    company: ICompany;
    hideDropdown: any;
}
