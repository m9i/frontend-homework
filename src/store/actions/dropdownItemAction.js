import * as actionTypes from "./actionTypes";

export const dropdownItemAction = (item) => ({
  type: actionTypes.SET_COMPANY,
  payload: item,
});
