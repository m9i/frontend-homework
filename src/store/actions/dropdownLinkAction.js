import * as actionTypes from "./actionTypes";

export const dropdownLinkAction = () => ({
  type: actionTypes.SHOW_DROPDOWN,
  showDropdown: false,
});
