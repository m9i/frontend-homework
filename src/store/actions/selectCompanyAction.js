import * as actionTypes from "./actionTypes";

export const selectCompanyAction = (selectedCompany) => ({
  type: actionTypes.SET_COMPANY,
  selectedCompany,
});
