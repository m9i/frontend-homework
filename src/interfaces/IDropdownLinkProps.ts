import {IToggleDropdownAction} from '../store/actions/toggleDropdownAction';

export interface IDropdownLinkProps {
    toggle: boolean;
    showDropdown: (() => (IToggleDropdownAction));
    hideDropdown: (() => (IToggleDropdownAction));
    selectedCompany: any;
}
