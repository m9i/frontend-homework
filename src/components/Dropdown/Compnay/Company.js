import React from "react"
import { selectCompanyAction } from "../../../store/actions/selectCompanyAction"
import { connect } from "react-redux"
import { toggleDropdownAction } from "../../../store/actions/toggleDropdownAction"

const Company = ({ onCompanyChange, selectedCompany, company, hideDropdown }) => {
  const onCompanyChanged = () => {
    onCompanyChange(company)
    hideDropdown()
  }

  return (
    <>
      <div
        className={
          company.value === selectedCompany.value ? "selected-company" : null
        }
        onClick={() => onCompanyChanged()}
      >
        <span> {company.value}</span>
        {company.value === selectedCompany.value && (
          <i className="material-icons-outlined nav__list__item--icon">done</i>
        )}
      </div>
    </>
  )
}

const mapStateToProps = (state) => ({
  selectedCompany: state.company.selectedCompany,
})

const mapDispatchToProps = (dispatch) => ({
  onCompanyChange: (item) => dispatch(selectCompanyAction(item)),
  hideDropdown: () => dispatch(toggleDropdownAction(false)),
})

export default connect(mapStateToProps, mapDispatchToProps)(Company)
