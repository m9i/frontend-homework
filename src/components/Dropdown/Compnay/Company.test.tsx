import React from 'react';
import Company from './Company';
import {mount, ReactWrapper, shallow, ShallowWrapper} from 'enzyme';
import {Provider} from 'react-redux';
import configureMockStore from 'redux-mock-store';
import {ICompany} from '../../../interfaces/ICompany';
import {ISelectCompanyAction} from '../../../store/actions/selectCompanyAction';

describe('Company', () => {
    const mockStore = configureMockStore([]);
    let store: any;
    let companyInfo: ICompany;

    beforeEach(() => {
        companyInfo = {id: 123, value: 'mock-name'};
        store = mockStore({
            dropdown: {},
            company: {
                selectedCompany: {value: 'mock-name'},
            },
        });
    });

    it('snapshot', () => {
        const CompanyComponent: ShallowWrapper = shallow(
            <Provider store={store}>
                <Company company={companyInfo}/>
            </Provider>
        ).dive();

        expect(CompanyComponent).toMatchSnapshot();
    });

    it('should show company name', () => {
        const CompanyComponent: ReactWrapper = mount(
            <Provider store={store}>
                <Company company={companyInfo}/>
            </Provider>
        );
        const companyName: ReactWrapper = CompanyComponent.find('span');
        expect(companyName.text()).toContain('mock-name');
    });

    it('should set "selected-company" css class for selected company', () => {
        const CompanyComponent: ReactWrapper = mount(
            <Provider store={store}>
                <Company company={companyInfo}/>
            </Provider>
        );
        const wrapper: ReactWrapper = CompanyComponent.find('div');

        expect(wrapper.hasClass('selected-company')).toEqual(true);
    });

    it('should show check icon when company has selected', () => {
        const CompanyComponent: ReactWrapper = mount(
            <Provider store={store}>
                <Company company={companyInfo}/>
            </Provider>
        );
        const icon: ReactWrapper = CompanyComponent.find('i');

        expect(icon.exists()).toEqual(true);
    });

    it('should hideDropdown when a company clicked', () => {
        const spyHideDropdown = jest.spyOn(store, 'dispatch');
        const CompanyComponent: ReactWrapper = mount(
            <Provider store={store}>
                <Company company={companyInfo}/>
            </Provider>
        );
        const expectedDispatch: ISelectCompanyAction = {
            selectedCompany: {id: 123, value: 'mock-name'},
            type: 'SET_COMPANY',
        };
        const wrapper: ReactWrapper = CompanyComponent.find('div');
        wrapper.simulate('click');

        expect(spyHideDropdown).toHaveBeenCalledTimes(2);
        expect(spyHideDropdown).toHaveBeenCalledWith(expectedDispatch);
    });
});
