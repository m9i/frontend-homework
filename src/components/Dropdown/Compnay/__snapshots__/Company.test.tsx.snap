// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Company snapshot 1`] = `
<Connect(Company)
  company={
    Object {
      "id": 123,
      "value": "mock-name",
    }
  }
/>
`;
